### PR TITLE
Partition: add key-partition document

### PR DESCRIPTION
**Pull request by qiancai**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions](https://github.com/pingcap/docs/blob/master/CONTRIBUTING.md#guideline-for-choosing-the-affected-versions).

- [x] master (the latest development version)
- [x] v7.0 (TiDB 7.0 versions)
- [ ] v6.6 (TiDB 6.6 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.4 (TiDB 6.4 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs-cn/pull/13223
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: partitioned-table.md [hunk 1]
```diff
@@ -560,6 +561,88 @@ MOD(YEAR('2005-09-01'),4)
...
+
+Both Key partitioning and Hash partitioning can evenly distribute data into a certain number of partitions. The difference is that Hash partitioning only supports distributing data based on a specified integer expression or an integer column, while Key partitioning supports distributing data based on a column list, and partitioning columns of Key partitioning are not limited to the integer type. The Hash algorithm of TiDB for Key partitioning is different from that of MySQL, so the table data distribution is also different.
+
+Partitioning by Key requires you to append a `PARTITION BY Key (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num`, where `num` is a positive integer indicating how many partitions a table is divided into.
+
+The following operation creates a Key partitioned table, which is divided into 4 partitions by `store_id`:
+
+{{< copyable "sql" >}}
+
```
Comment (qiancai) [inline partitioned-table.md:575]:
```suggestion
```

File: partitioned-table.md [hunk 2]
```diff
@@ -560,6 +561,88 @@ MOD(YEAR('2005-09-01'),4)
...
+PARTITION BY KEY(store_id)
+PARTITIONS 4;
+```
+
+If `PARTITIONS num` is not specified, the default number of partitions is 1.
+
+You can also create a Key partitioned table based on non-integer columns such as VARCHAR. For example, you can partition a table by the `fname` column:
+
+{{< copyable "sql" >}}
+
```
Comment (qiancai) [inline partitioned-table.md:595]:
```suggestion
```

File: partitioned-table.md [hunk 3]
```diff
@@ -560,6 +561,88 @@ MOD(YEAR('2005-09-01'),4)
...
+    store_id INT
+)
+PARTITION BY KEY(fname)
+PARTITIONS 4;
+```
+
+You can also create a Key partitioned table based on multiple columns. For example, you can divide a table into 4 partitions based on `fname` and `store_id`:
+
+{{< copyable "sql" >}}
+
```
Comment (qiancai) [inline partitioned-table.md:613]:
```suggestion
```

File: partitioned-table.md [hunk 4]
```diff
@@ -560,6 +561,88 @@ MOD(YEAR('2005-09-01'),4)
...
+    store_id INT
+)
+PARTITION BY KEY(fname, store_id)
+PARTITIONS 4;
+```
+
+Currently, TiDB does not support creating Key partitioned tables if the partition column list specified in `PARTITION BY KEY` is empty. For example, after you execute the following statement, TiDB will create a non-partitioned table and return an `Unsupported partition type KEY, treat as normal table` warning.
+
+{{< copyable "sql" >}}
+
```
Comment (qiancai) [inline partitioned-table.md:631]:
```suggestion
```

**Review by TonsnakeLin — APPROVED** [review]

File: partitioned-table.md [hunk 5]
```diff
@@ -841,6 +926,20 @@ alter table members optimize partition p0;
 ERROR 8200 (HY000): Unsupported optimize partition
 ```
 
+### Key partition management
+
+Currently, Key partitioning only supports the `ALTER TABLE ... TRUNCATE PARTITION` partition management statement.
+
+If you execute a Key partition management statement that is not yet supported, TiDB will return an error.
```
Comment (lilin90) [inline partitioned-table.md:933]: Please limit the use of "will" and try to use the present tense.
```suggestion
If you execute a Key partition management statement that is not yet supported, TiDB returns an error.
```

File: partitioned-table.md [hunk 6]
```diff
@@ -560,6 +561,80 @@ MOD(YEAR('2005-09-01'),4)
 =  1
 ```
 
+### Key partitioning
+
+Starting from v7.0.0, TiDB supports Key partitioning. For TiDB versions earlier than v7.0.0, if you try creating a Key partitioned table, TiDB will create it as a non-partitioned table and return a warning.
```
Comment (qiancai) [inline partitioned-table.md:566]:
```suggestion
Starting from v7.0.0, TiDB supports Key partitioning. For TiDB versions earlier than v7.0.0, if you try creating a Key partitioned table, TiDB creates it as a non-partitioned table and returns a warning.
```

File: partitioned-table.md [hunk 7]
```diff
@@ -570,6 +645,12 @@ Since v6.4.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR HASH` syntax
 
 - For all other statements of MySQL Linear Hash partitions, they work in TiDB the same as that in MySQL, except that the rows are distributed differently if the number of partitions is not a power of 2, which will give different results for [partition selection](#partition-selection), `TRUNCATE PARTITION`, and `EXCHANGE PARTITION`.
 
+### How TiDB handles Linear Key partitions
+
+Starting from v7.0.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR HASH` syntax for Key partitioning. However, TiDB ignores the `LINEAR` keyword and uses a non-linear hash algorithm instead.
+
+Before v7.0.0, if you try creating a Key partitioned table, TiDB will create it as a non-partitioned table and return a warning.
```
Comment (qiancai) [inline partitioned-table.md:652]:
```suggestion
Before v7.0.0, if you try creating a Key partitioned table, TiDB creates it as a non-partitioned table and returns a warning.
```

File: partitioned-table.md (hunk 5 above)
Comment (qiancai) [reply partitioned-table.md]: Thanks for the comments. The other places are updated in https://github.com/pingcap/docs/pull/12963/commits/00def2a14f5a99297b89e783528c880f01e40c21.

File: partitioned-table.md [hunk 8]
```diff
@@ -841,6 +926,20 @@ alter table members optimize partition p0;
 ERROR 8200 (HY000): Unsupported optimize partition
 ```
 
+### Key partition management
+
+Currently, Key partitioning only supports the `ALTER TABLE ... TRUNCATE PARTITION` partition management statement.
+
```
Comment (qiancai) [inline partitioned-table.md:932]:
````suggestion
```sql
ALTER TABLE members TRUNCATE PARTITION p0;
```

```
Query OK, 0 rows affected (0.03 sec)
````
```

File: partitioned-table.md [hunk 9]
```diff
@@ -10,9 +10,10 @@ This document introduces TiDB's implementation of partitioning.
...
+This section introduces the types of partitioning in TiDB. Currently, TiDB supports [Range partitioning](#range-partitioning), [Range COLUMNS partitioning](#range-columns-partitioning), [List partitioning](#list-partitioning), [List COLUMNS partitioning](#list-columns-partitioning), [Hash partitioning](#hash-partitioning), and [Key partitioning](#key-partitioning).
 
-Range partitioning, Range COLUMNS partitioning, List partitioning and List COLUMNS partitioning are used to resolve the performance issues caused by a large amount of deletions in the application, and support fast drop partition operations. Hash partitioning is used to scatter the data when there are a large amount of writes.
+- Range partitioning, Range COLUMNS partitioning, List partitioning, and List COLUMNS partitioning are used to resolve the performance issues caused by a large number of deletions in the application, and support fast drop partition operations.
```
Comment (qiancai) [inline partitioned-table.md:15]:
```suggestion
- Range partitioning, Range COLUMNS partitioning, List partitioning, and List COLUMNS partitioning are used to resolve the performance issues caused by a large number of deletions in the application, and support dropping partitions quickly.
```

File: partitioned-table.md [hunk 10]
```diff
@@ -560,6 +561,80 @@ MOD(YEAR('2005-09-01'),4)
...
+```sql
+CREATE TABLE employees (
+    id INT NOT NULL,
+    fname VARCHAR(30),
+    lname VARCHAR(30),
+    hired DATE NOT NULL DEFAULT '1970-01-01',
+    separated DATE DEFAULT '9999-12-31',
+    job_code INT,
+    store_id INT
+)
```
Comment (qiancai) [inline partitioned-table.md:583]:
```suggestion
)

```

File: partitioned-table.md [hunk 11]
```diff
@@ -560,6 +561,80 @@ MOD(YEAR('2005-09-01'),4)
...
```
Comment (qiancai) [inline partitioned-table.md:601]:
```suggestion
)

```

File: partitioned-table.md [hunk 12]
```diff
@@ -1350,7 +1457,9 @@ YEARWEEK()
 
 ### Compatibility with MySQL
 
-Currently, TiDB supports Range partitioning, Range COLUMNS partitioning, List partitioning, List COLUMNS partitioning, and Hash partitioning. Other partitioning types that are available in MySQL such as key partitioning are not supported yet in TiDB.
+Currently, TiDB supports Range partitioning, Range COLUMNS partitioning, List partitioning, List COLUMNS partitioning, Hash partitioning, and Key partitioning. Other partitioning types that are available in MySQL are not supported yet in TiDB.
```
[the remark this answers is not in the document]
Comment (qiancai) [reply partitioned-table.md:1460]: Good catch. Added a comment in https://github.com/pingcap/docs-cn/pull/13223/files#r1141761876

[the remark this answers is not in the document]
Comment (qiancai) [reply partitioned-table.md:1464]: 中文已更新 https://github.com/pingcap/docs-cn/pull/13223/commits/ee043bc87f85da621ab974669b63005f0751ec1d

**Review by hfxsd — APPROVED** [review]

File: mysql-compatibility.md [hunk 13]
```diff
@@ -148,7 +148,7 @@ In TiDB, all supported DDL changes are performed online. Compared with DDL opera
 * The `ALGORITHM={INSTANT,INPLACE,COPY}` syntax functions only as an assertion in TiDB, and does not modify the `ALTER` algorithm. See [`ALTER TABLE`](/sql-statements/sql-statement-alter-table.md) for further details.
 * Adding/Dropping the primary key of the `CLUSTERED` type is unsupported. For more details about the primary key of the `CLUSTERED` type, refer to [clustered index](/clustered-indexes.md).
 * Different types of indexes (`HASH|BTREE|RTREE|FULLTEXT`) are not supported, and will be parsed and ignored when specified.
-* Table Partitioning supports `HASH`, `RANGE`, and `LIST` partitioning types. For the unsupported partition type, the `Warning: Unsupported partition type %s, treat as normal table` error might be output, where `%s` is a specific partition type.
+* Table Partitioning supports `HASH`, `RANGE`, `LIST`, and `KEY` partitioning types. Currently, the `KEY` partition type does not support partition statements with an empty partition column list. For an unsupported partition type, TiDB might return `Warning: Unsupported partition type %s, treat as normal table`, where `%s` is the specific partition type.
```
Comment (mjonss) [inline mysql-compatibility.md:151]:
```suggestion
* Table Partitioning supports `HASH`, `RANGE`, `LIST`, and `KEY` partitioning types. Currently, the `KEY` partition type does not support partition statements with an empty partition column list. For an unsupported partition type, TiDB will return `Warning: Unsupported partition type %s, treat as normal table`, where `%s` is the specific partition type.
```

File: partitioned-table.md [hunk 14]
```diff
@@ -560,6 +561,84 @@ MOD(YEAR('2005-09-01'),4)
...
+
+Both Key partitioning and Hash partitioning can evenly distribute data into a certain number of partitions. The difference is that Hash partitioning only supports distributing data based on a specified integer expression or an integer column, while Key partitioning supports distributing data based on a column list, and partitioning columns of Key partitioning are not limited to the integer type. The Hash algorithm of TiDB for Key partitioning is different from that of MySQL, so the table data distribution is also different.
+
+Partitioning by Key requires you to append a `PARTITION BY Key (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num`, where `num` is a positive integer indicating how many partitions a table is divided into.
```
Comment (mjonss) [inline partitioned-table.md:570]:
```suggestion
Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you also need to append `PARTITIONS num`, where `num` is a positive integer indicating how many partitions a table is divided into, or named definitions of the partitions like `(PARTITION pOne, PARTITION pTwo).
```

**Review by mjonss — APPROVED** [review]
Some minor suggestions

File: partitioned-table.md (hunk 14 above)
Comment (qiancai) [reply partitioned-table.md:570]: @TonsnakeLin PTAL

File: partitioned-table.md [hunk 15]
```diff
@@ -570,6 +649,12 @@ Since v6.4.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR HASH` syntax
 
 - For all other statements of MySQL Linear Hash partitions, they work in TiDB the same as that in MySQL, except that the rows are distributed differently if the number of partitions is not a power of 2, which will give different results for [partition selection](#partition-selection), `TRUNCATE PARTITION`, and `EXCHANGE PARTITION`.
 
+### How TiDB handles Linear Key partitions
+
+Starting from v7.0.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR HASH` syntax for Key partitioning. However, TiDB ignores the `LINEAR` keyword and uses a non-linear hash algorithm instead.
```
Comment (qiancai) [inline partitioned-table.md:654]:
```suggestion
Starting from v7.0.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR KEY` syntax for Key partitioning. However, TiDB ignores the `LINEAR` keyword and uses a non-linear hash algorithm instead.
```

File: partitioned-table.md [hunk 16]
```diff
@@ -560,6 +561,84 @@ MOD(YEAR('2005-09-01'),4)
...
+
+Both Key partitioning and Hash partitioning can evenly distribute data into a certain number of partitions. The difference is that Hash partitioning only supports distributing data based on a specified integer expression or an integer column, while Key partitioning supports distributing data based on a column list, and partitioning columns of Key partitioning are not limited to the integer type. The Hash algorithm of TiDB for Key partitioning is different from that of MySQL, so the table data distribution is also different.
+
+Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you also need to append `PARTITIONS num`, where `num` is a positive integer indicating how many partitions a table is divided into, or named definitions of the partitions like `(PARTITION pOne, PARTITION pTwo).
```
Comment (qiancai) [inline partitioned-table.md:570]:
```suggestion
Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num` (where `num` is a positive integer indicating how many partitions a table is divided into), or add the definition of the partition names, for example, adding `(PARTITION p0, PARTITION p1)` means dividing the table into two partitions, with partition names `p0` and `p1`.
```

File: partitioned-table.md [hunk 17]
```diff
@@ -560,6 +561,84 @@ MOD(YEAR('2005-09-01'),4)
...
+
+Both Key partitioning and Hash partitioning can evenly distribute data into a certain number of partitions. The difference is that Hash partitioning only supports distributing data based on a specified integer expression or an integer column, while Key partitioning supports distributing data based on a column list, and partitioning columns of Key partitioning are not limited to the integer type. The Hash algorithm of TiDB for Key partitioning is different from that of MySQL, so the table data distribution is also different.
+
+Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num` (where `num` is a positive integer indicating how many partitions a table is divided into), or add the definition of the partition names (for example, adding `(PARTITION p0, PARTITION p1)` means dividing the table into two partitions, with partition names `p0` and `p1`).
```
Comment (qiancai) [inline partitioned-table.md:570]:
```suggestion
Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num` (where `num` is a positive integer indicating how many partitions a table is divided into), or append the definition of the partition names (for example, adding `(PARTITION p0, PARTITION p1)` means dividing the table into two partitions, with partition names `p0` and `p1`).
```

File: partitioned-table.md [hunk 18]
```diff
@@ -560,6 +561,84 @@ MOD(YEAR('2005-09-01'),4)
...
+
+Both Key partitioning and Hash partitioning can evenly distribute data into a certain number of partitions. The difference is that Hash partitioning only supports distributing data based on a specified integer expression or an integer column, while Key partitioning supports distributing data based on a column list, and partitioning columns of Key partitioning are not limited to the integer type. The Hash algorithm of TiDB for Key partitioning is different from that of MySQL, so the table data distribution is also different.
+
+Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num` (where `num` is a positive integer indicating how many partitions a table is divided into), or append the definition of the partition names (for example, adding `(PARTITION p0, PARTITION p1)` means dividing the table into two partitions, with partition names `p0` and `p1`).
```
Comment (hfxsd) [inline partitioned-table.md:570]:
```suggestion
When partitioning by Key, you need to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support `GEOMETRY`). In addition, you might also need to append `PARTITIONS num` (where `num` is a positive integer indicating how many partitions a table is divided into), or append the definition of the partition names (for example, adding `(PARTITION p0, PARTITION p1)` means dividing the table into two partitions named `p0` and `p1`.
```

**Pull request merged**